<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

# Project 1: Data Analysis of Singapore's Rainfall And Its' Influence On Dengue Cases

--- 
# Part 1: Background & Data Cleaning


---

### Contents:
- [Background](#Background)
- [Data Import & Cleaning](#Data-Import-and-Cleaning)

## Background

Singapore's climate does not vary much from month to month, and there are no distinct wet or dry seasons.<sup>[1](http://www.weather.gov.sg/climate-climate-of-singapore/#:~:text=Singapore%20is%20situated%20near%20the,month%2Dto%2Dmonth%20variation)</sup> However, a higher rainfall occurs during the two monsoon seasons from November to January and June to September.

Its' metrics such as temperature, humidity, sunshine, and rainfall etc. have been recorded by the government at the Changi climate station.

Dengue fever (DF) and dengue hemorrhagic fever (DHF) are endemic diseases in many regions. Transmission of the virus is made through the female *Aedes aegypti* mosquito.<sup>[2](https://www.who.int/news-room/fact-sheets/detail/dengue-and-severe-dengue)</sup>

Singapore, along with many other tropical countries, take the brunt of the dengue virus<sup>[3](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3373041/)</sup> due to the countries' climates being ideal breeding grounds for Aedes aegypti. DF and DHF is an increasing concern for many in Singapore, especially during the Covid-19 pandemic period.<sup>[4](https://www.straitstimes.com/singapore/health/more-people-have-died-from-dengue-than-from-covid-19-in-singapore-this-year)</sup> As dengue cases are on the rise, it is important to look at how the spread of the disease can be minimised.

## Problem Statement

As such, we want to <font color = 'green'>**predict future dengue cases using the weather data in Singapore**</font> so appropriate measures can be put in place to minimise the number of dengue cases.

### Outside Research

[1] [Meteorological Service Singapore](http://www.weather.gov.sg/climate-climate-of-singapore/#:~:text=Singapore%20is%20situated%20near%20the,month%2Dto%2Dmonth%20variation) <br/>
[2] [World Health Organization](https://www.who.int/news-room/fact-sheets/detail/dengue-and-severe-dengue) <br/>
[3] [Dengue Prevention and 35 Years of Vector Control in Singapore](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3373041/) <br/>
[4] [2020 Covid and Dengue Rise](https://www.straitstimes.com/singapore/health/more-people-have-died-from-dengue-than-from-covid-19-in-singapore-this-year) <br/>
[5] [2021 Highest Dengue Cases In 4 Months](https://www.straitstimes.com/singapore/dengue-cases-cross-6000-mark-in-less-than-4-months-exceeding-whole-of-2021) <br/>
[6] [National Centre for Infectious Diseases](https://www.ncid.sg/Health-Professionals/Articles/Pages/Epidemic-Dengue-in-Singapore-During-COVID-19-Pandemic.aspx) <br/>
[7] [Effects of Temperature and Heatwaves on Dengue in Singapore](https://www.sciencedirect.com/science/article/pii/S0048969721001832#:~:text=Studies%20have%20shown%20that%20an,et%20al.%2C%202009)<br/>
[8] [Highest Dengue Death Tolls](https://www.straitstimes.com/singapore/health/singapore-records-19-dengue-deaths-in-2022-nearly-four-times-2021-s-toll)

### Data Sources

* [Singapore's Monthly Weather Data – singstat.gov.sg](https://tablebuilder.singstat.gov.sg/table/TS/M890081)
* [Singapore's Weekly Infectious Diseases – data.gov.sg](https://data.gov.sg/dataset/weekly-infectious-disease-bulletin-cases)



In [2]:
# Importing all libraries used: 

import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns

## Data Import and Cleaning

### Weather Data

#### Importing and examining weather data

In [3]:
# Importing main monthly dataframe with rain, humidity, etc.
weather = pd.read_csv('../data/monthly_nea_singstat_rain_humidity.csv')
weather.head()

Data Series  2023 May   2023 Apr   \
0  Air Temperature Means Daily Maximum (Degree Ce...       32.9       32.6   
1  Air Temperature Means Daily Minimum (Degree Ce...       26.7       25.8   
2  Air Temperature Absolute Extremes Maximum (Deg...       35.0       34.3   
3  Air Temperature Absolute Extremes Minimum (Deg...       23.4       23.0   
4                        Total Rainfall (Millimetre)      112.8      222.8   

   2023 Mar   2023 Feb   2023 Jan   2022 Dec   2022 Nov   2022 Oct   \
0       30.9       31.0       30.1       30.3       31.4       31.5   
1       24.6       24.5       24.5       24.5       24.4       24.7   
2       33.6       32.0       34.2       32.7       33.9       34.6   
3       21.8       21.7       22.4       23.1       22.4       23.1   
4      243.4      324.4      302.6      215.4      313.8      279.0   

   2022 Sep   ...  1975 Oct   1975 Sep   1975 Aug   1975 Jul   1975 Jun   \
0       31.1  ...       30.5       30.3       30.3         30       30.6   
1       25.4  ...       23.3       23.5       23.6       22.9       23.7   
2       32.6  ...       32.6       32.1       31.7       31.7       32.2   
3       22.4  ...       21.5       21.7       20.5       20.3       21.5   
4      121.0  ...      169.8      146.5        177      192.5      212.8   

   1975 May   1975 Apr   1975 Mar   1975 Feb   1975 Jan   
0       31.4       31.1       31.1       30.4       30.5  
1         24       23.8       23.1         23       22.8  
2       33.1       32.8       32.5       32.9       31.8  
3       22.2       22.1       21.8       21.4         20  
4      104.6       79.7      197.9       98.6        121  

[5 rows x 582 columns]

In [4]:
#Checking shape and info
print(weather.shape)
weather.info()

(10, 582)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Columns: 582 entries, Data Series to 1975 Jan 
dtypes: float64(473), object(109)
memory usage: 45.6+ KB


We will need to transpose the data to have more accurate stats

In [5]:
# Transposing data
weather_m = weather.T
weather_m.head()

0  \
Data Series  Air Temperature Means Daily Maximum (Degree Ce...   
2023 May                                                  32.9   
2023 Apr                                                  32.6   
2023 Mar                                                  30.9   
2023 Feb                                                  31.0   

                                                             1  \
Data Series  Air Temperature Means Daily Minimum (Degree Ce...   
2023 May                                                  26.7   
2023 Apr                                                  25.8   
2023 Mar                                                  24.6   
2023 Feb                                                  24.5   

                                                             2  \
Data Series  Air Temperature Absolute Extremes Maximum (Deg...   
2023 May                                                  35.0   
2023 Apr                                                  34.3   
2023 Mar                                                  33.6   
2023 Feb                                                  32.0   

                                                             3  \
Data Series  Air Temperature Absolute Extremes Minimum (Deg...   
2023 May                                                  23.4   
2023 Apr                                                  23.0   
2023 Mar                                                  21.8   
2023 Feb                                                  21.7   

                                       4  \
Data Series  Total Rainfall (Millimetre)   
2023 May                           112.8   
2023 Apr                           222.8   
2023 Mar                           243.4   
2023 Feb                           324.4   

                                                     5  \
Data Series  Highest Daily Rainfall Total (Millimetre)   
2023 May                                          34.6   
2023 Apr                                          85.8   
2023 Mar                                          90.2   
2023 Feb                                         154.8   

                                         6                                  7  \
Data Series  Number Of Rainy Days (Number)  Bright Sunshine Daily Mean (Hour)   
2023 May                              11.0                                6.6   
2023 Apr                              19.0                                5.8   
2023 Mar                              12.0                                6.0   
2023 Feb                              13.0                                5.1   

                                                8  \
Data Series  Minimum Relative Humidity (Per Cent)   
2023 May                                     43.0   
2023 Apr                                     45.0   
2023 Mar                                     52.0   
2023 Feb                                     50.0   

                                                      9  
Data Series  24 Hours Mean Relative Humidity (Per Cent)  
2023 May                                           73.7  
2023 Apr                                           77.4  
2023 Mar                                           79.2  
2023 Feb                                           79.4

#### Renaming and resetting index of weather data

In [6]:
# Renaming the column
weather_m.columns = weather_m.values[0]
weather_m.drop('Data Series', axis=0, inplace=True)
weather_m.head()

Air Temperature Means Daily Maximum (Degree Celsius)  \
2023 May                                                32.9     
2023 Apr                                                32.6     
2023 Mar                                                30.9     
2023 Feb                                                31.0     
2023 Jan                                                30.1     

          Air Temperature Means Daily Minimum (Degree Celsius)  \
2023 May                                                26.7     
2023 Apr                                                25.8     
2023 Mar                                                24.6     
2023 Feb                                                24.5     
2023 Jan                                                24.5     

          Air Temperature Absolute Extremes Maximum (Degree Celsius)  \
2023 May                                                35.0           
2023 Apr                                                34.3           
2023 Mar                                                33.6           
2023 Feb                                                32.0           
2023 Jan                                                34.2           

          Air Temperature Absolute Extremes Minimum (Degree Celsius)  \
2023 May                                                23.4           
2023 Apr                                                23.0           
2023 Mar                                                21.8           
2023 Feb                                                21.7           
2023 Jan                                                22.4           

          Total Rainfall (Millimetre)  \
2023 May                        112.8   
2023 Apr                        222.8   
2023 Mar                        243.4   
2023 Feb                        324.4   
2023 Jan                        302.6   

          Highest Daily Rainfall Total (Millimetre)  \
2023 May                                       34.6   
2023 Apr                                       85.8   
2023 Mar                                       90.2   
2023 Feb                                      154.8   
2023 Jan                                       72.0   

          Number Of Rainy Days (Number) Bright Sunshine Daily Mean (Hour)  \
2023 May                           11.0                               6.6   
2023 Apr                           19.0                               5.8   
2023 Mar                           12.0                               6.0   
2023 Feb                           13.0                               5.1   
2023 Jan                           22.0                               4.4   

          Minimum Relative Humidity (Per Cent)  \
2023 May                                  43.0   
2023 Apr                                  45.0   
2023 Mar                                  52.0   
2023 Feb                                  50.0   
2023 Jan                                  43.0   

          24 Hours Mean Relative Humidity (Per Cent)  
2023 May                                        73.7  
2023 Apr                                        77.4  
2023 Mar                                        79.2  
2023 Feb                                        79.4  
2023 Jan                                        81.9

In [7]:
# Reindexing
weather_m.reset_index(names='date',inplace=True)
weather_m.head()

date Air Temperature Means Daily Maximum (Degree Celsius)  \
0  2023 May                                                32.9     
1  2023 Apr                                                32.6     
2  2023 Mar                                                30.9     
3  2023 Feb                                                31.0     
4  2023 Jan                                                30.1     

  Air Temperature Means Daily Minimum (Degree Celsius)  \
0                                               26.7     
1                                               25.8     
2                                               24.6     
3                                               24.5     
4                                               24.5     

  Air Temperature Absolute Extremes Maximum (Degree Celsius)  \
0                                               35.0           
1                                               34.3           
2                                               33.6           
3                                               32.0           
4                                               34.2           

  Air Temperature Absolute Extremes Minimum (Degree Celsius)  \
0                                               23.4           
1                                               23.0           
2                                               21.8           
3                                               21.7           
4                                               22.4           

  Total Rainfall (Millimetre) Highest Daily Rainfall Total (Millimetre)  \
0                       112.8                                      34.6   
1                       222.8                                      85.8   
2                       243.4                                      90.2   
3                       324.4                                     154.8   
4                       302.6                                      72.0   

  Number Of Rainy Days (Number) Bright Sunshine Daily Mean (Hour)  \
0                          11.0                               6.6   
1                          19.0                               5.8   
2                          12.0                               6.0   
3                          13.0                               5.1   
4                          22.0                               4.4   

  Minimum Relative Humidity (Per Cent)  \
0                                 43.0   
1                                 45.0   
2                                 52.0   
3                                 50.0   
4                                 43.0   

  24 Hours Mean Relative Humidity (Per Cent)  
0                                       73.7  
1                                       77.4  
2                                       79.2  
3                                       79.4  
4                                       81.9

In [8]:
# Renaming the column names
weather_m.columns = ['date', 'daily_max_temp', 'daily_min_temp', 
                     'extreme_max_temp', 'extreme_min_temp', 
                     'total_rainfall', 'total_highest_daily_rainfall',
                     'no_of_rainy_days', 'sunshine_daily_mean',
                     'min_relative_humidity', 'mean_daily_humidity']
weather_m.head()

date daily_max_temp daily_min_temp extreme_max_temp extreme_min_temp  \
0  2023 May            32.9           26.7             35.0             23.4   
1  2023 Apr            32.6           25.8             34.3             23.0   
2  2023 Mar            30.9           24.6             33.6             21.8   
3  2023 Feb            31.0           24.5             32.0             21.7   
4  2023 Jan            30.1           24.5             34.2             22.4   

  total_rainfall total_highest_daily_rainfall no_of_rainy_days  \
0          112.8                         34.6             11.0   
1          222.8                         85.8             19.0   
2          243.4                         90.2             12.0   
3          324.4                        154.8             13.0   
4          302.6                         72.0             22.0   

  sunshine_daily_mean min_relative_humidity mean_daily_humidity  
0                 6.6                  43.0                73.7  
1                 5.8                  45.0                77.4  
2                 6.0                  52.0                79.2  
3                 5.1                  50.0                79.4  
4                 4.4                  43.0                81.9

#### Converting the data to their appropriate types

In [9]:
# Re-examining shape and info
print(weather_m.shape)
weather_m.info()

(581, 11)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 581 entries, 0 to 580
Data columns (total 11 columns):
 #   Column                        Non-Null Count  Dtype 
---  ------                        --------------  ----- 
 0   date                          581 non-null    object
 1   daily_max_temp                581 non-null    object
 2   daily_min_temp                581 non-null    object
 3   extreme_max_temp              581 non-null    object
 4   extreme_min_temp              581 non-null    object
 5   total_rainfall                581 non-null    object
 6   total_highest_daily_rainfall  581 non-null    object
 7   no_of_rainy_days              581 non-null    object
 8   sunshine_daily_mean           581 non-null    object
 9   min_relative_humidity         581 non-null    object
 10  mean_daily_humidity           581 non-null    object
dtypes: object(11)
memory usage: 50.1+ KB


In [10]:
# defining function to convert to numeric
def num(df,column) :
    df[column]=pd.to_numeric(df[column])

In [11]:
# Converting all except date to numeric
num(weather_m, 'daily_max_temp')
num(weather_m, 'daily_min_temp')
num(weather_m, 'extreme_max_temp')
num(weather_m, 'extreme_min_temp')
num(weather_m, 'total_rainfall')
num(weather_m, 'total_highest_daily_rainfall')
num(weather_m, 'no_of_rainy_days')
num(weather_m, 'sunshine_daily_mean')
# num(weather_m, 'min_relative_humidity')
num(weather_m, 'mean_daily_humidity')

In [12]:
weather_m.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 581 entries, 0 to 580
Data columns (total 11 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   date                          581 non-null    object 
 1   daily_max_temp                581 non-null    float64
 2   daily_min_temp                581 non-null    float64
 3   extreme_max_temp              581 non-null    float64
 4   extreme_min_temp              581 non-null    float64
 5   total_rainfall                581 non-null    float64
 6   total_highest_daily_rainfall  581 non-null    float64
 7   no_of_rainy_days              581 non-null    float64
 8   sunshine_daily_mean           581 non-null    float64
 9   min_relative_humidity         581 non-null    object 
 10  mean_daily_humidity           581 non-null    float64
dtypes: float64(9), object(2)
memory usage: 50.1+ KB


Upon attempting to change object into float, realise min_relative_humidity has na values. Checking and converting these na to np.nan values.

In [13]:
# Investigating Minimum Relative Humidity as object...
weather_m['min_relative_humidity'].unique()

# Found that there are na values.

array([43.0, 45.0, 52.0, 50.0, 46.0, 47.0, 53.0, 54.0, 41.0, 51.0, 48.0,
       49.0, 42.0, 55.0, 44.0, 38.0, 37.0, 35.0, 40.0, 33.0, 39.0, 60.0,
       57.0, 59.0, 34.0, 36.0, 61.0, 58.0, 56.0, 'na'], dtype=object)

In [14]:
# Replacing na with np.nan
weather_m['min_relative_humidity'].replace('na', np.nan, inplace=True)
weather_m[weather_m['min_relative_humidity'].isna()].head()
weather_m['min_relative_humidity'] = pd.to_numeric(weather_m['min_relative_humidity'])

In [15]:
# Checking info data again...
weather_m.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 581 entries, 0 to 580
Data columns (total 11 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   date                          581 non-null    object 
 1   daily_max_temp                581 non-null    float64
 2   daily_min_temp                581 non-null    float64
 3   extreme_max_temp              581 non-null    float64
 4   extreme_min_temp              581 non-null    float64
 5   total_rainfall                581 non-null    float64
 6   total_highest_daily_rainfall  581 non-null    float64
 7   no_of_rainy_days              581 non-null    float64
 8   sunshine_daily_mean           581 non-null    float64
 9   min_relative_humidity         473 non-null    float64
 10  mean_daily_humidity           581 non-null    float64
dtypes: float64(10), object(1)
memory usage: 50.1+ KB


#### Adding appropriate information to convert date type

In [16]:
# Now to format Data Series into month and year.
# Adding a substitute day first (1st day of month)
weather_m['date'] = weather_m['date'] + '01'

In [17]:
# Converting to datetime format
weather_m['date'] = pd.to_datetime(weather_m['date'], format= '%Y %b %d')

In [18]:
# Converting month & year to two separate columns
weather_m.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 581 entries, 0 to 580
Data columns (total 11 columns):
 #   Column                        Non-Null Count  Dtype         
---  ------                        --------------  -----         
 0   date                          581 non-null    datetime64[ns]
 1   daily_max_temp                581 non-null    float64       
 2   daily_min_temp                581 non-null    float64       
 3   extreme_max_temp              581 non-null    float64       
 4   extreme_min_temp              581 non-null    float64       
 5   total_rainfall                581 non-null    float64       
 6   total_highest_daily_rainfall  581 non-null    float64       
 7   no_of_rainy_days              581 non-null    float64       
 8   sunshine_daily_mean           581 non-null    float64       
 9   min_relative_humidity         473 non-null    float64       
 10  mean_daily_humidity           581 non-null    float64       
dtypes: datetime64[ns](1), float64(10

In [19]:
weather_m.head()

date  daily_max_temp  daily_min_temp  extreme_max_temp  \
0 2023-05-01            32.9            26.7              35.0   
1 2023-04-01            32.6            25.8              34.3   
2 2023-03-01            30.9            24.6              33.6   
3 2023-02-01            31.0            24.5              32.0   
4 2023-01-01            30.1            24.5              34.2   

   extreme_min_temp  total_rainfall  total_highest_daily_rainfall  \
0              23.4           112.8                          34.6   
1              23.0           222.8                          85.8   
2              21.8           243.4                          90.2   
3              21.7           324.4                         154.8   
4              22.4           302.6                          72.0   

   no_of_rainy_days  sunshine_daily_mean  min_relative_humidity  \
0              11.0                  6.6                   43.0   
1              19.0                  5.8                   45.0   
2              12.0                  6.0                   52.0   
3              13.0                  5.1                   50.0   
4              22.0                  4.4                   43.0   

   mean_daily_humidity  
0                 73.7  
1                 77.4  
2                 79.2  
3                 79.4  
4                 81.9

### Monthly Mean Temp Data

In [20]:
#Importing mean temp data
temp_mean = pd.read_csv('../data/monthly_mean_temp.csv')
temp_mean.info()
temp_mean.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 497 entries, 0 to 496
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   month      497 non-null    object 
 1   mean_temp  497 non-null    float64
dtypes: float64(1), object(1)
memory usage: 7.9+ KB


month  mean_temp
0  1982-01       25.9
1  1982-02       27.1
2  1982-03       27.2
3  1982-04       27.0
4  1982-05       28.0

#### Converting to proper formats

In [21]:
# converting month to proper format
temp_mean['month'] = temp_mean['month'] + '-01'

In [22]:
temp_mean['month'] = pd.to_datetime(temp_mean['month'], format= '%Y-%m-%d')
temp_mean.head()

month  mean_temp
0 1982-01-01       25.9
1 1982-02-01       27.1
2 1982-03-01       27.2
3 1982-04-01       27.0
4 1982-05-01       28.0

#### Renaming columns

In [23]:
temp_mean.columns = ['date', 'monthly_mean_temp']
temp_mean.head()

date  monthly_mean_temp
0 1982-01-01               25.9
1 1982-02-01               27.1
2 1982-03-01               27.2
3 1982-04-01               27.0
4 1982-05-01               28.0

### Dengue Data

#### Importing and examining dengue data

In [24]:
# Importing weekly disease numbers 
disease_w = pd.read_csv('../data/weekly-infectious-disease-bulletin-cases.csv')
disease_w.head()
disease_w.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20070 entries, 0 to 20069
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   epi_week      20070 non-null  object
 1   disease       20070 non-null  object
 2   no._of_cases  20070 non-null  int64 
dtypes: int64(1), object(2)
memory usage: 470.5+ KB


#### Dropping unnecessary rows 
We only need the dengue data from this dataframe

In [25]:
# Replacing dataframe with rows that only contain 'Dengue' in the 'disease' column
disease_w = disease_w[disease_w['disease'].str.contains('Dengue')]
disease_w.reset_index(drop=True, inplace=True)
disease_w.head()

epi_week                    disease  no._of_cases
0  2012-W01               Dengue Fever            74
1  2012-W01  Dengue Haemorrhagic Fever             0
2  2012-W02               Dengue Fever            64
3  2012-W02  Dengue Haemorrhagic Fever             2
4  2012-W03               Dengue Fever            60

#### Converting date into proper format

In [26]:
# Adding start of week into epi_week column
disease_w['epi_week'] = disease_w['epi_week'].str.replace('W','')
disease_w['epi_week'] = disease_w['epi_week'] + '-Sun'

In [27]:
# Converting to datetime format
disease_w.head()
disease_w['epi_week'] = pd.to_datetime(disease_w['epi_week'], format= '%Y-%U-%a')

In [28]:
# Checking formats
print(disease_w.info())
disease_w.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1148 entries, 0 to 1147
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   epi_week      1148 non-null   datetime64[ns]
 1   disease       1148 non-null   object        
 2   no._of_cases  1148 non-null   int64         
dtypes: datetime64[ns](1), int64(1), object(1)
memory usage: 27.0+ KB
None


epi_week                    disease  no._of_cases
0 2012-01-01               Dengue Fever            74
1 2012-01-01  Dengue Haemorrhagic Fever             0
2 2012-01-08               Dengue Fever            64
3 2012-01-08  Dengue Haemorrhagic Fever             2
4 2012-01-15               Dengue Fever            60

## Combining Data

Combining weekly disease information to crete a new dataframe with combined monthly dengue data so we can combine it with weather data


In [29]:
disease_w.info()
weather_m.info()
temp_mean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1148 entries, 0 to 1147
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   epi_week      1148 non-null   datetime64[ns]
 1   disease       1148 non-null   object        
 2   no._of_cases  1148 non-null   int64         
dtypes: datetime64[ns](1), int64(1), object(1)
memory usage: 27.0+ KB
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 581 entries, 0 to 580
Data columns (total 11 columns):
 #   Column                        Non-Null Count  Dtype         
---  ------                        --------------  -----         
 0   date                          581 non-null    datetime64[ns]
 1   daily_max_temp                581 non-null    float64       
 2   daily_min_temp                581 non-null    float64       
 3   extreme_max_temp              581 non-null    float64       
 4   extreme_min_temp              581 non-null    float64       
 5   to

### Merging weather and mean temperature

In [30]:
weather_m = weather_m.merge(temp_mean, how='left', on='date')

In [31]:
weather_m.head()

date  daily_max_temp  daily_min_temp  extreme_max_temp  \
0 2023-05-01            32.9            26.7              35.0   
1 2023-04-01            32.6            25.8              34.3   
2 2023-03-01            30.9            24.6              33.6   
3 2023-02-01            31.0            24.5              32.0   
4 2023-01-01            30.1            24.5              34.2   

   extreme_min_temp  total_rainfall  total_highest_daily_rainfall  \
0              23.4           112.8                          34.6   
1              23.0           222.8                          85.8   
2              21.8           243.4                          90.2   
3              21.7           324.4                         154.8   
4              22.4           302.6                          72.0   

   no_of_rainy_days  sunshine_daily_mean  min_relative_humidity  \
0              11.0                  6.6                   43.0   
1              19.0                  5.8                   45.0   
2              12.0                  6.0                   52.0   
3              13.0                  5.1                   50.0   
4              22.0                  4.4                   43.0   

   mean_daily_humidity  monthly_mean_temp  
0                 73.7               29.5  
1                 77.4               28.5  
2                 79.2               27.1  
3                 79.4               26.9  
4                 81.9               26.5

### Merging weather and disease data

#### Examining more before merging

In [32]:
print(disease_w['epi_week'].describe(datetime_is_numeric=True))
# Earliest dengue date starts from 2012 Jan, ends on 2022 Dec

count                             1148
mean     2017-07-01 16:40:58.536585216
min                2012-01-01 00:00:00
25%                2014-10-05 00:00:00
50%                2017-06-28 12:00:00
75%                2020-04-05 00:00:00
max                2022-12-25 00:00:00
Name: epi_week, dtype: object


#### Earliest dengue date starts from 2012 Jan, ends on 2022 Dec

splitting up dengue and dhf numbers to combine monthly data

In [33]:
disease_m = disease_w.set_index('epi_week')
disease_m_dengue = disease_m[disease_m['disease'] == 'Dengue Fever']
disease_m_dengue = disease_m_dengue.resample('M').sum(numeric_only = True)

disease_m_dhf = disease_m[disease_m['disease'] == 'Dengue Haemorrhagic Fever']
disease_m_dhf = disease_m_dhf.resample('M').sum(numeric_only = True)

print(disease_m_dengue.head())
print(disease_m_dhf.head())

disease_m = disease_m_dengue.join(disease_m_dhf, on='epi_week', lsuffix = 'dengue_fever', rsuffix= 'dhf').reset_index()
disease_m.head()

            no._of_cases
epi_week                
2012-01-31           332
2012-02-29           257
2012-03-31           229
2012-04-30           360
2012-05-31           337
            no._of_cases
epi_week                
2012-01-31             6
2012-02-29             1
2012-03-31             2
2012-04-30             3
2012-05-31             1


epi_week  no._of_casesdengue_fever  no._of_casesdhf
0 2012-01-31                       332                6
1 2012-02-29                       257                1
2 2012-03-31                       229                2
3 2012-04-30                       360                3
4 2012-05-31                       337                1

#### Renaming columns of disease dataframe

In [34]:
disease_m.columns = 'date', 'dengue_fever_cases', 'dhf_cases'
disease_m['date'].describe(datetime_is_numeric=True)

count                              132
mean     2017-07-15 19:16:21.818181888
min                2012-01-31 00:00:00
25%                2014-10-23 06:00:00
50%                2017-07-15 12:00:00
75%                2020-04-07 12:00:00
max                2022-12-31 00:00:00
Name: date, dtype: object

#### Filtering Weather data to 2012 to 2022 so dataframes can be combined

In [35]:
weather_filtered = weather_m[(weather_m['date'] > '2011-12-31') & (weather_m['date'] < '2023-01-01')]
weather_filtered.sort_values('date', ascending=True, inplace=True)
weather_filtered.reset_index(drop = True, inplace=True)
weather_filtered.head()

/var/folders/_s/4bjxzfvd5jj9yc3x2ssfcc0h0000gn/T/ipykernel_62943/1191973572.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  weather_filtered.sort_values('date', ascending=True, inplace=True)


date  daily_max_temp  daily_min_temp  extreme_max_temp  \
0 2012-01-01            30.6            24.3              32.5   
1 2012-02-01            31.1            24.8              32.9   
2 2012-03-01            31.1            24.4              33.2   
3 2012-04-01            31.4            24.8              32.9   
4 2012-05-01            31.7            24.8              33.7   

   extreme_min_temp  total_rainfall  total_highest_daily_rainfall  \
0              21.5           106.1                          26.2   
1              22.7            83.6                          38.4   
2              21.1           313.4                          92.3   
3              21.9           260.6                          48.4   
4              22.0           292.0                          79.2   

   no_of_rainy_days  sunshine_daily_mean  min_relative_humidity  \
0              16.0                  6.0                   59.0   
1              11.0                  7.0                   55.0   
2              21.0                  4.2                   58.0   
3              20.0                  5.8                   54.0   
4              19.0                  5.6                   52.0   

   mean_daily_humidity  monthly_mean_temp  
0                 87.3               26.7  
1                 85.6               27.2  
2                 87.3               27.0  
3                 85.1               27.6  
4                 83.1               28.2

#### Checking the data of each dataframe

In [36]:
weather_filtered.info()
disease_m.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 132 entries, 0 to 131
Data columns (total 12 columns):
 #   Column                        Non-Null Count  Dtype         
---  ------                        --------------  -----         
 0   date                          132 non-null    datetime64[ns]
 1   daily_max_temp                132 non-null    float64       
 2   daily_min_temp                132 non-null    float64       
 3   extreme_max_temp              132 non-null    float64       
 4   extreme_min_temp              132 non-null    float64       
 5   total_rainfall                132 non-null    float64       
 6   total_highest_daily_rainfall  132 non-null    float64       
 7   no_of_rainy_days              132 non-null    float64       
 8   sunshine_daily_mean           132 non-null    float64       
 9   min_relative_humidity         132 non-null    float64       
 10  mean_daily_humidity           132 non-null    float64       
 11  monthly_mean_temp             13

Both dataframes have 132 rows.

Now we are making sure the order of the date is same as we want to keep the date from the disease dataframe as it represents the last day of each month instead of the first.

In [37]:
# checking to see the order of date in disease dataframe is the same

disease_m.sort_values('date', ascending=True, inplace= True)
disease_m.head()

date  dengue_fever_cases  dhf_cases
0 2012-01-31                 332          6
1 2012-02-29                 257          1
2 2012-03-31                 229          2
3 2012-04-30                 360          3
4 2012-05-31                 337          1

Combining both data as both are the same number or rows and have been correctly re-ordered

In [38]:
# Combining the two dataframes together

weather_disease = disease_m.join(weather_filtered, how = 'left', rsuffix='_w')
weather_disease.drop('date_w', axis=1, inplace= True)
weather_disease.head()

date  dengue_fever_cases  dhf_cases  daily_max_temp  daily_min_temp  \
0 2012-01-31                 332          6            30.6            24.3   
1 2012-02-29                 257          1            31.1            24.8   
2 2012-03-31                 229          2            31.1            24.4   
3 2012-04-30                 360          3            31.4            24.8   
4 2012-05-31                 337          1            31.7            24.8   

   extreme_max_temp  extreme_min_temp  total_rainfall  \
0              32.5              21.5           106.1   
1              32.9              22.7            83.6   
2              33.2              21.1           313.4   
3              32.9              21.9           260.6   
4              33.7              22.0           292.0   

   total_highest_daily_rainfall  no_of_rainy_days  sunshine_daily_mean  \
0                          26.2              16.0                  6.0   
1                          38.4              11.0                  7.0   
2                          92.3              21.0                  4.2   
3                          48.4              20.0                  5.8   
4                          79.2              19.0                  5.6   

   min_relative_humidity  mean_daily_humidity  monthly_mean_temp  
0                   59.0                 87.3               26.7  
1                   55.0                 85.6               27.2  
2                   58.0                 87.3               27.0  
3                   54.0                 85.1               27.6  
4                   52.0                 83.1               28.2

In [39]:
weather_disease.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 132 entries, 0 to 131
Data columns (total 14 columns):
 #   Column                        Non-Null Count  Dtype         
---  ------                        --------------  -----         
 0   date                          132 non-null    datetime64[ns]
 1   dengue_fever_cases            132 non-null    int64         
 2   dhf_cases                     132 non-null    int64         
 3   daily_max_temp                132 non-null    float64       
 4   daily_min_temp                132 non-null    float64       
 5   extreme_max_temp              132 non-null    float64       
 6   extreme_min_temp              132 non-null    float64       
 7   total_rainfall                132 non-null    float64       
 8   total_highest_daily_rainfall  132 non-null    float64       
 9   no_of_rainy_days              132 non-null    float64       
 10  sunshine_daily_mean           132 non-null    float64       
 11  min_relative_humidity         13

## Exporting combined data into a csv

In [41]:
weather_disease.to_csv('../data/weather_disease_monthly_data.csv', index = False) # dataframe_to_export.to_csv('path/filename.format')

---

# Data Dictionary

----

*weather_disease_monthly_data.csv*

|Feature|Type|Datasource|Description|
|---|---|---|---|
|dengue_fever_cases|integer|weekly-infectious-disease-bulletin-cases|Number of recorded cases of Dengue Fever in an epidemiological week|
|dhf_cases|integer|weekly-infectious-disease-bulletin-cases|Number of recorded cases of Dengue Haemorrhagic Fever in an epidemiological week|
|daily_max_temp|float|monthly_nea_singstat_rain_humidity|The monthly and annual mean daily minimum temperature recorded at the Changi Climate Station|
|daily_min_temp|float|monthly_nea_singstat_rain_humidity|The monthly and annual mean daily maximum temperature recorded at the Changi Climate Station|
|extreme_max_temp|float|monthly_nea_singstat_rain_humidity|The monthly extreme maximum air temperature recorded at the Changi Climate Station|
|extreme_min_temp|float|monthly_nea_singstat_rain_humidity|The absolute extreme minimum air temperature recorded at the Changi Climate Station|
|total_rainfall|float|monthly_nea_singstat_rain_humidity|The total monthly rainfall recorded at the Changi Climate Station|
|total_highest_daily_rainfall|float|monthly_nea_singstat_rain_humidity|The highest daily total rainfall for the month recorded at the Changi Climate Station|
|no_of_rainy_days|float|monthly_nea_singstat_rain_humidity|The number of rain days* in a month recorded at the Changi Climate Station|
|sunshine_daily_mean|float|monthly_nea_singstat_rain_humidity|The monthly mean sunshine hours in a day recorded at the Changi Climate Station|
|min_relative_humidity|float|monthly_nea_singstat_rain_humidity|The absolute extreme minimum relative humidity for the month recorded at the Changi Climate Station|
|mean_daily_humidity|float|monthly_nea_singstat_rain_humidity|The monthly mean relative humidity recorded at the Changi Climate Station|
|monthly_mean_temp|float|monthly_nea_singstat_rain_humidity|The monthly mean air temperature recorded at the Changi Climate Station|

*\* day with rainfall amount of 0.2mm or more*